src/metabase/models/collection.clj

```diff
@@ -731,69 +799,101 @@
          :location [:like (str (children-location collection) "%")]
          additional-conditions))
 
-(mu/defn archive-or-unarchive-collection!
-  "Archive or un-archive a collection, moving it to the trash if necessary. Note that namespaced collections are never
-  moved to the trash."
+(mu/defn archive-collection!
+  "Mark a collection as archived, along with all its children."
+  [collection :- CollectionWithLocationAndIDOrRoot]
+  (api/check-403
+   (perms/set-has-full-permissions-for-set?
+    @api/*current-user-permissions-set*
+    (perms-for-archiving collection)))
+  (let [trash-operation-id      (str (random-uuid))
+        affected-collection-ids (cons (u/the-id collection)
+                                      (collection->descendant-ids collection
+                                                                  :archived [:not= true]))]
+    (t2/with-transaction [_conn]
+      (t2/update! :model/Collection (u/the-id collection)
+                  {:trash_operation_id trash-operation-id
+                   :trashed_directly   true
+                   :archived           true})
+      (t2/query-one
+       {:update :collection
+        :set    {:trash_operation_id trash-operation-id
+                 :trashed_directly   false
+                 :archived           true}
+        :where  [:and
+                 [:like :location (str (children-location collection) "%")]
+                 [:not :archived]]})
+      (doseq [model [:model/NativeQuerySnippet :model/Pulse :model/Timeline]]
+        (t2/update! model {:collection_id [:in affected-collection-ids]}
+                    {:archived true}))
+      (doseq [model [:model/Card :model/Dashboard]]
```

Comment:
There are a couple places now in this namespace where we enumerate the "models which can live in a collection" which makes me worried that it'll be easy to miss updating these if a new model is added. 

Maybe this could be pulled out as a standalone var? I kind of like the concept of `collectable-models`. Then you'd only need to hard-code Card and Dashboard in these functions as the ones that can be trashed directly (or that could be a separate var)